pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator.java

```diff
@@ -0,0 +1,193 @@
+        return DirectiveDescriptor.all();
+    }
+
+    @Restricted(DoNotUse.class) // accessed via REST API
+    public HttpResponse doGenerateDirective(StaplerRequest req, @QueryParameter String json) throws Exception {
+        // TODO is there not an easier way to do this? Maybe Descriptor.newInstancesFromHeteroList on a one-element JSONArray?
+        JSONObject jsonO = JSONObject.fromObject(json);
+        Jenkins j = Jenkins.getActiveInstance();
+        Class<?> c = j.getPluginManager().uberClassLoader.loadClass(jsonO.getString("stapler-class"));
+        DirectiveDescriptor descriptor = (DirectiveDescriptor)j.getDescriptor(c.asSubclass(AbstractDirective.class));
+        if (descriptor == null) {
+            return HttpResponses.plainText("<could not find " + c.getName() + ">");
+        }
+        Object o;
+        try {
+            o = descriptor.newInstance(req, jsonO);
+        } catch (RuntimeException x) { // e.g. IllegalArgumentException
+            return HttpResponses.plainText(Functions.printThrowable(x));
+        }
+        try {
+            String groovy = descriptor.toIndentedGroovy((AbstractDirective)o);
+            return HttpResponses.plainText(groovy);
+        } catch (UnsupportedOperationException x) {
+            LOGGER.log(Level.WARNING, "failed to render " + json, x);
+            return HttpResponses.plainText(x.getMessage());
+        }
+    }
+
+    @Restricted(NoExternalUse.class)
+    public static String mapToClosure(Map<String,?> args) {
+        StringBuilder result = new StringBuilder("{\n");
+        for (Map.Entry<String,?> arg : args.entrySet()) {
+            if (!(arg.getValue() instanceof String && arg.getValue().equals(""))) {
+                result.append(arg.getKey()).append(" ");
+                if (arg.getValue() instanceof Map) {
+                    result.append(mapToClosure((Map<String, ?>) arg.getValue()));
+                } else if (arg.getValue() instanceof UninstantiatedDescribable) {
+                    result.append(mapToClosure(((UninstantiatedDescribable) arg.getValue()).getArguments()));
+                } else if (arg.getValue() != null) {
+                    result.append(Snippetizer.object2Groovy(arg.getValue())).append("\n");
+                }
+            }
+        }
+        result.append("}\n");
+        return result.toString();
+    }
+
+    @Restricted(NoExternalUse.class)  // For jelly and internal use
+    public static String getSymbolForDescriptor(Descriptor d) {
+        if (d instanceof StepDescriptor) {
+            return ((StepDescriptor) d).getFunctionName();
+        } else {
+            Set<String> symbols = SymbolLookup.getSymbolValue(d);
+            if (!symbols.isEmpty()) {
+                return symbols.iterator().next();
+            } else {
+                return "(unknown)";
+            }
+        }
+    }
+
+    @Restricted(DoNotUse.class)
+    @Extension public static class PerJobAdder extends TransientActionFactory<Job> {
```

Comment:
Shouldn't this be removed?

Reply:
Annoyingly, no - get rid of that and `someJob/directive-generator` stops working.